File: Design-Patterns-Using-JavaScript/Basics OOPS/Inheritance.js
class Person {
    constructor(name, age, gender) {
        this.name = name;
        this.age = age;
        this.gender = gender;
    }

    toString() {
        return `Person: ${this.name} is a ${this.gender} and ${this.age} years old.`;
    }

    getGender() {
        return this.gender;
    }
}

class Citizen extends Person {
    constructor(name, age, id, gender) {
        super(name, age, gender);
        this.id = id;
    }

    getVoterId() {
        return this.id;
    }

    toString() {
        return `Citizen: ${this.name} is a ${this.gender} and ${this.age} years old with voter id ${this.id}.`;
    }
}

// Client code
const p = new Person("John", 32, "Male");
console.log(p.toString());

const c = new Citizen("Smith", 31, 1234, "Male");
console.log(c.toString());

/*
Person: John is a Male and 32 years old.
Citizen: Smith is a Male and 31 years old with voter id 1234.
*/

File: Design-Patterns-Using-Swift/Basics OOPS/Inheritance.swift
class Person {
    var name: String
    var age: Int
    var gender: String

    init(name: String, age: Int, gender: String) {
        self.name = name
        self.age = age
        self.gender = gender
    }

    var description: String {
        return "Person: \(name) is a \(gender) and \(age) years old."
    }

    func getGender() -> String {
        return gender
    }
}

class Citizen: Person {
    private var id: Int

    init(name: String, age: Int, id: Int, gender: String) {
        self.id = id
        super.init(name: name, age: age, gender: gender)
    }

    func getVoterId() -> Int {
        return id
    }

    override var description: String {
        return "Citizen: \(name) is a \(gender) and \(age) years old with voter id \(id)."
    }
}

let p = Person(name: "John", age: 32, gender: "Male")
print(p.description)

let c = Citizen(name: "Smith", age: 31, id: 1234, gender: "Male")
print(c.description)
